proof_ics23.go

```diff
@@ -0,0 +1,170 @@
+package iavl
+
+import (
+	"bytes"
+	"fmt"
+
+	ics23 "github.com/confio/ics23/go"
+	"github.com/tendermint/go-amino"
+)
+
+/*
+GetMembershipProof will produce a CommitmentProof that the given key (and queries value) exists in the iavl tree.
+If the key doesn't exist in the tree, this will return an error.
+*/
+func (t *ImmutableTree) GetMembershipProof(key []byte) (*ics23.CommitmentProof, error) {
+	exist, err := createExistenceProof(t, key)
+	if err != nil {
+		return nil, err
+	}
+	proof := &ics23.CommitmentProof{
+		Proof: &ics23.CommitmentProof_Exist{
+			Exist: exist,
+		},
+	}
+	return proof, nil
+}
+
+/*
+GetNonMembershipProof will produce a CommitmentProof that the given key doesn't exist in the iavl tree.
+If the key exists in the tree, this will return an error.
+*/
+func (t *ImmutableTree) GetNonMembershipProof(key []byte) (*ics23.CommitmentProof, error) {
+	// idx is one node right of what we want....
+	var err error
+	idx, val := t.GetWithIndex(key)
+
+	if val != nil {
+		return nil, fmt.Errorf("cannot create NonExistanceProof when Key in State")
+	}
+
+	nonexist := &ics23.NonExistenceProof{
+		Key: key,
+	}
+
+	if idx >= 1 {
+		leftkey, _ := t.GetByIndex(idx - 1)
+		if err != nil {
+			return nil, err
+		}
+
+		nonexist.Left, err = createExistenceProof(t, leftkey)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// this will be nil if nothing right of the queried key
+	rightkey, _ := t.GetByIndex(idx)
+	if err != nil {
+		return nil, err
+	}
+
+	if rightkey != nil {
+		nonexist.Right, err = createExistenceProof(t, rightkey)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	proof := &ics23.CommitmentProof{
+		Proof: &ics23.CommitmentProof_Nonexist{
+			Nonexist: nonexist,
+		},
+	}
+	return proof, nil
+}
+
+func createExistenceProof(tree *ImmutableTree, key []byte) (*ics23.ExistenceProof, error) {
```

Comment:
The cosmos make `createExistenceProof` an interface of `ImmutableTree`, any reason we change it?